File: 78.subsets/subsets.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int> > res;
        vector<int> path;
        dfs(nums, path, 0, res);
        return res;
    }
    void dfs(vector<int>& nums, vector<int>& path, int start, vector<vector<int> >& res) {
        if (start == nums.size()) {
            res.push_back(path);
            return;
        }
        dfs (nums, path, start + 1, res);
        path.push_back(nums[start]);
        dfs (nums, path, start + 1, res);
        path.pop_back();
    }
};

int main()
{
    Solution s;
    int a[3] = {1,2,3};
    vector<int> vec(a, a+3);
    vector<vector<int> > res = s.subsets(vec);
    vector<vector<int> >::iterator it = res.begin();
    while (it != res.end()) {
        for (int i : *it)
            cout << i << " ";
        cout << endl;
        ++it;
    }
    return 0;
}

File: 78.subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        dfs(res, path, nums, 0);
        return res;
    }
    
    void dfs(List<List<Integer>> res, List<Integer> path, int[] nums, int index) {
        res.add(new ArrayList<>(path));
        for (int i = index; i < nums.length; i++) {
            path.add(nums[i]);
            dfs(res, path, nums, i + 1);
            path.remove(path.size() - 1);
        }
    }
}
